SUP-117-Transform-Data-Audit (#409)
Macrometacorp/docs: docs/cep/tutorials/transforming-data.md

## Introduction

The stream processor allows you to perform a wide range of transformations to the input data received. The transformations are carried out via operators that are defined inline within the stream worker.

## Transform data using Operators

> The operators that you can configure inline within Stream workers in order to carry out data transformations are listed in the [Stream Query Guide](../query-guide/index.md).

Reviewer: ```suggestion
The operators that you can configure inline within stream workers in order to carry out data transformations are listed in the [Stream Query Guide](../query-guide/index.md).
```

To show how an inline operators are configured, let's consider an example where readings from a sensor that indicates 
the temperature of a room every second are transformed to indicate the average temperature and the average humidity as at each second.

> 1. Open the GUI. Click the **Stream Apps** tab.

Reviewer: ```suggestion
1. Log in to the Macrometa web console. Click the **Stream Apps** tab.
```

    
        ```sql
        SELECT roomNo, avg(temp) AS avgTemp
        ```

    C. Add the `from` clause with the name of the input stream to indicate that the events to be processed are taken from the input stream.

        ```sql
        FROM TempStream
        ```

    D. To group by a specific attribute (by the `roomNo` attribute in this example), specify it via the `group by` clause as shown below.

        ```sql
        INSERT INTO OutputStream
        SELECT roomNo, avg(temp) AS avgTemp
        FROM TempStream
        GROUP BY roomNo;
        ```

9. Save the stream worker. The completed stream worker is as follows.

    ```sql
    @App:name("TemperatureApp")
    @App:description("Calculate an average temperature of the room")
    @App:qlVersion("2")
    
    CREATE STREAM TempStream (deviceID long, roomNo int, temp double);
    
    CREATE SINK OutputStream WITH (type='stream', stream='OutputStream', map.type='json') (roomNo int, avgTemp double);
    
    INSERT INTO OutputStream
    SELECT roomNo, avg(temp) AS avgTemp
    FROM TempStream
    GROUP BY roomNo;
    ```

## Transform Data Using Plugins

> The stream processor offers a variety of options to carry out data transformations via in-built plugins. The following table describes the complete list of plugins that provide data transformation functionality.

Reviewer: By stream processor, do you mean stream worker?

Author: I think stream processor, 

Stream Processor

This allows events to be collected, generated, dropped and expired anytime by altering the event format by adding one or more attributes to it based on the given input parameters.

> 5. Add the following sample stream application.

Reviewer: In all cases, change "stream application" to "stream worker."

They used to call stream workers "stream applications." This is now outdated terminology, like calling Macrometa C8DB.

> 6. Let's define the input stream to define the schema based on which data is selected to the streaming integration flow.

Reviewer: ```suggestion
6. Let's define the input stream to define the schema based on which data are selected to the streaming integration flow.
```

>     In this example, let's assume that each event indicates the device ID, the room number, and the temperature. Therefore, let's define an input stream as follows:

Reviewer: Reading through this set of instructions, it seems like almost every use of "let's" is unnecessary. Try deleting them and see how it reads.

    ```sql
	   CREATE STREAM TempStream (deviceID long, roomNo int, temp double);
    ```
       
   > For more information about defining input streams to receive events, see the [Consuming Data page](consuming-data.md).
           
7. Let's define the output stream which will receive the average temperature of each incomming message in the TempStream. The output stream definition is as follows:
    ```sql
       CREATE SINK OutputStream WITH (type='stream', stream='OutputStream', map.type='json') (roomNo int, avgTemp double);
    ```

8. To do the required transformation, let's add the query as follows:

>     1. Add the `INSERT INTO` clause with the name of the output stream to indicate that the processed events are directed to that stream.

Reviewer: Do not use numbers as a sublist, use letters.

    
        ```sql
        SELECT roomNo, avg(temp) AS avgTemp
        ```

    1. Add the `from` clause with the name of the input stream to indicate that the events to be processed are taken from the input stream.

        ```sql
        FROM TempStream
        ```

    1. To group by a specific attribute (by the `roomNo` attribute in this example), specify it via the `group by` clause as shown below.

        ```sql
        INSERT INTO OutputStream
        SELECT roomNo, avg(temp) AS avgTemp
        FROM TempStream
        GROUP BY roomNo;
        ```

9. Save the stream application. The completed stream application is as follows.

    ```sql
    @App:name("TemperatureApp")
    @App:description("Calculate an average temperature of the room")
    @App:qlVersion("2")
    
    CREATE STREAM TempStream (deviceID long, roomNo int, temp double);
    
    CREATE SINK OutputStream WITH (type='stream', stream='OutputStream', map.type='json') (roomNo int, avgTemp double);
    
    INSERT INTO OutputStream
    SELECT roomNo, avg(temp) AS avgTemp
    FROM TempStream
    GROUP BY roomNo;
    ```

> ## Transform data using Plugins

Reviewer: ```suggestion
## Transform Data Using Plugins
```
